src/odemis/model/_vattributes.py

```diff
@@ -982,6 +983,30 @@ def _check(self, value):
         Enumerated._check(self, value)
         VigilantAttribute._check(self, value)
 
+    def clip(self, val):
+        """ Clip the given value to fit within the choices.
+
+        If the range contains vectors of length n, each element of val will be
+        clipped separately by position for positions 0..n-1
+        return (same type as val): value clipped
+        """
+
```

Comment:
Probably, for "safety", you could first check whether the proposed value is one of the choices, and in such case, short-cut to immediately accept (aka return) it.

Then, I think you should handle the case where the choices are just simple numbers:
* The choices are all numbers or are tuple of numbers: -> return the closest choices from the given value, using the euclidean distance

The trickiest part is to know whether all the choices are tuples of numbers. It might be possible to just check whether the proposed value (whether it's a tuple and all the elements are numbers), then compare to the choices for which the `distance.euclidean` returns a value (and not fail).